Lib/asyncio/locks.py

```diff
@@ -378,20 +383,18 @@ async def acquire(self):
         fut = self._get_loop().create_future()
         self._waiters.append(fut)
 
-        # Finally block should be called before the CancelledError
-        # handling as we don't want CancelledError to call
-        # _wake_up_first() and attempt to wake up itself.
         try:
-            try:
-                await fut
-            finally:
-                self._waiters.remove(fut)
-        except exceptions.CancelledError:
-            if not fut.cancelled():
+            await fut
+        except BaseException:
+            self._waiters.remove(fut)
+            if fut.done() and not fut.cancelled():
```

Comment:
Oooh, interesting! Here you are checking `fut.done()` -- in the corresponding code in `Lock` you aren't. What's the difference?

Reply:
Ah, the Semaphore.  (fut.done() and not fut.cancelled()) means that the `release()` and `_wake_up_next()` successfully notified *this* Task.  Only later did it become cancelled, and so, we must rebalance the books (undo our wakeup, and do another .   The old code always _assumes_ that the Future is `done()` because it _either_ got `set_result()`  _or_ `cancel()`, (via Task.cancel()).  But there is theoretical third case:  The Task gets some other error causing it to stop waiting for the future, without a `Future.set_result()` or Future.cancel()`, such as being the recipient of a `send_exception()`.

Without playing special favorites to my library code which does in fact send tasks direct interrupts and can wake up a Task without touching the Future it is waiting for, I think it makes sense to test precisely for the _exact_ case where a successful `Semaphore.release()` was believed to have been handed to a Task.

The contrast with Lock() is that there is only one Task which can hold the lock at any one time, and no need for bookkeeping.  The wakeup function always wakes up the task at the head of the queue, possibly multiple times.  In the absence of an exception, a Task which wakes up in the Lock would always find `self._locked == False` (and could probably assert that).  And it will proceed to set it to True.   And if a waking task is _not_ going to take the lock, it is safe to wake up any sleeping task _at the head_ of the queue.

Reply:
Wow. The code was carefully redone 14 months ago and until a minute ago I believed it was correct. But it looks like, indeed, if the future was not yet completed when the task was cancelled (this can happen through a timeout too), we'd be waking up another task even though we weren't given a "token" yet.

Interestingly, in this case we abandon the future without it being done (by removing it from the list of waiters).

Reply:
Well, `Task.cancel()` actually has the property of also cancelling any `Future` which the task is awaiting.  This makes sense in the context of the original design of `cancel()` which is _supposed_ to propagate to any `Task`s being awaited.    But when just _interrupting_ a Task, i.e. asking it to _abandon_ its await, the logic breaks down.

The change to the Semaphore was not intended to be a big one, just _affirming_ the implied "Future.done()` with an actual test.  But there is also something else incorrect there, the final wakeup in the successful `acquire()` case should never be need.  I think I will add assertions to this code to illustrate my point, we can remove them later if you like